### PR TITLE
Fix python decomps for OpOverloadPackets and add tests

### LINKED ISSUE #107713: Ensure python decompositions registered for OpOverloadPacket's work for out overloads
### 🐛 Describe the bug

If a python decomposition is registered for an `OpOverloadPacket`, care must be taken to ensure the decomposition functions for each `OpOverload`, particularly for operators that have `out` variants. For example the `OpOverloadPacket` `torch.ops.aten.smooth_l1_loss` has an  `out` overload `torch.ops.aten.smooth_l1_loss.out` so its python decomposition should support receiving `out` arguments. 

For example callling the below `out` overload of the `smooth_l1_loss` op will currently result in error for the `Meta` dispatch key since its python decomposition does not receive `out` arguments.

```c++
at::_ops::smooth_l1_loss_out::redispatch(self, target, reduction, beta, out);
```

```python

@register_decomposition(aten.smooth_l1_loss)
@pw_cast_for_opmath
def smooth_l1_loss(..):
    ....
```

For torch ops that support the `out` parameter, `test_meta_outplace` and `test_dispatch_meta_outplace` cover checking python decompositions still function. However torch ops may not have `out` parameters but still have `out` aten overloads that are called e.g. via `structured_delegate`. 

### Versions

n/a

cc @ezyang @eellison @bdhirsh @albanD @SherlockNoMad

### PR DESCRIPTION
**Pull request by kundaMwiza**
- Extend `test_torch_dispatch_meta_outplace` to test torch ops that do not have an out parameter but have aten op overloads that have out parameters. Additionally, Python decompositions may register `OpOverloadPacket`'s so decompositions need to be tested to ensure all `OpOverloads` still function for the `Meta` key (e.g. if a python decomposition is registered for an aten op `aten.foo` with overloads `[default, out]`, the python function needs to support receiving out arguments)

- Add out parameter wrappers to python decomps for aten ops that have out overloads

CC. @ezyang @albanD @lezcano 

Fixes #107713

### CODE REVIEW

**Comment by ezyang** [comment]
@zou3519 letting you field this one

**Comment by zou3519** [comment]
@kundaMwiza  sorry, I'm behind on PR reviews. I'll be back next week to review this.

**Comment by zou3519** [comment]
@kundaMwiza I'm not very familiar with this code. Do you have a more detailed explanation of what test you're adding?

**Comment by kundaMwiza** [comment]
Hello, [this](https://github.com/pytorch/pytorch/issues/107713) may help

**Comment by kundaMwiza** [comment]
For torch ops that have an out parameter e.g. `torch.foo(..,out=out)`, the correct working of dispatching to the Meta key is tested in test_meta.py. In particular `def test_meta_outplace(..)` and `def test_dispatch_meta_outplace(...)` will ensure that both `torch.foo(out=None)` and `torch.foo(out=out_tensor)` work when the arguments are on the Meta device.

Now the problem is that some torch ops do not have an `out` parameter in their signature, but they do have an out ATen overload. For example `torch.smooth_l1_loss` does not have an out parameter but internally it is dispatched to `smooth_l1_loss.out` through a structured_delegate:

```yaml
- func: smooth_l1_loss.out(Tensor self, Tensor target, int reduction=Mean, float beta=1.0, *, Tensor(a!) out) -> Tensor(a!)
  device_check: NoCheck   # TensorIterator
  structured: True
  structured_inherits: TensorIteratorBase
  python_module: nn
  dispatch:
    CPU, CUDA: smooth_l1_loss_out
    MPS: smooth_l1_loss_out_mps

- func: smooth_l1_loss(Tensor self, Tensor target, int reduction=Mean, float beta=1.0) -> Tensor
  device_check: NoCheck   # TensorIterator
  structured_delegate: smooth_l1_loss.out <---
  python_module: nn
```

If a python decomposition is not registered for smooth_l1_loss, then ordinarily the meta implementation registered for `smooth_l1_loss.out` will be called, and so `def test_meta_outplace` in test_meta.py will test the correct working for Meta dispatch.

Now if a python decomposition is registered for all overloads of an op e.g. `smooth_l1_loss` and there is no _meta python registration_, then the python decomposition will be registered for Meta dispatch.

e.g.
```python
@register_decomposition("aten.smooth_l1_loss")
def smooth_l1_loss:
     ....
 ```

in _meta_registrations.py, all overloads of smooth_l1_loss will have the same python function for the Meta key~:

```python
torch.ops.aten.smooth_l1_loss.py_impl(torch._C.DispatchKey.Meta(smooth_l1_loss)
torch.ops.aten.smooth_l1_loss.out.py_impl(torch._C.DispatchKey.Meta(smooth_l1_loss)
```

Since a python decomposition is registered for `aten.smooth_l1_loss`, Meta dispatch will now call the python function `def `smooth_l1_loss` and will not call the structured delegate as above. And since `torch.smooth_l1_loss` does not have an out parameter, there is now no way of testing that the Meta dispatch for `aten.smooth_l1_loss.out` works (for example, the python decomposition may not have a signature that receives the `out` parameter and will therefore raise an exception at runtime).

The tests in `test_meta.py` are written to receive sample inputs from OpDB which provide arguments compatible with `torch.foo` functions. To test cases like `aten.smooth_l1_loss.out` explicitly for Meta dispatch with the same sample inputs as `aten.smooth_l1_loss`, it would be required to find a mapping between `torch.foo` and `aten.foo` which is not 1-1 from my trials.

The addition I have added in this PR is to extend the test `def test_dispatch_meta_outplace(...)`. Since the test uses `__torch_dispatch__`, it is possible to intercept the call to `aten.smooth_l1_loss`, call that with its inputs to get the output, and then also call `aten.smooth_l1_loss.out` with the same inputs to test that the `out` version also works for meta dispatch. This will ensure that if a python decomposition is added for an op with no `out` parameter in its corresponding torch op, the python decomposition will be tested to also work for the `out` version if the decomposition is also registered for that.

**Comment by zou3519** [comment]
@kundaMwiza  thanks for the detailed explanation. I have a couple more questions, let me put them inline

File: test/test_meta.py [hunk 1]
```diff
@@ -528,10 +527,12 @@ def run_meta_crossref(
             for a, ma in zip(flat_args, flat_meta_args):
                 flat_new_args.append(a if isinstance(a, torch.Tensor) and a.dtype in [torch.int8, torch.bool] else ma)
             meta_args = (meta_args[0], tree_unflatten(flat_new_args, spec))
-        elif func is torch.ops.aten.repeat_interleave.Tensor:
+        elif func in (torch.ops.aten.repeat_interleave.Tensor, torch.ops.aten.repeat_interleave.Tensor_out):
```
Comment (zou3519) [inline test/test_meta.py:530]: This function is `run_meta_crossref`. IIRC there is a test that runs meta tests over the entire PyTorch codebase, not just over OpInfos. Do your changes impact that test?

File: test/test_meta.py [hunk 2]
```diff
@@ -690,12 +691,6 @@ def run_meta_crossref(
 meta_function_device_expected_failures_only_outplace = defaultdict(dict)
 meta_function_device_skips = defaultdict(dict)
 
-meta_function_device_expected_failures['cpu'] = {
-    torch.native_batch_norm: {bf16},
-    torch._native_batch_norm_legit: {bf16},
-    torch.native_layer_norm: {bf16},
-}
-
```
Comment (zou3519) [inline test/test_meta.py:698]: Why were these removed?

Comment (kundaMwiza) [reply test/test_meta.py:698]: I can put the fixes in a different PR, I made the fixes for these in decompositions.py

File: test/test_meta.py (hunk 1 above)
Comment (kundaMwiza) [reply test/test_meta.py:530]: I'm only familiar with the test that runs over OpInfos, can you please share the test that runs meta tests over the entire codebase?

File: test/test_meta.py (hunk 2 above)
Comment (zou3519) [reply test/test_meta.py:698]: Yes, please separate out the fixes from the new tests. It makes it easier to tell what changed and why

File: test/test_meta.py (hunk 1 above)
Comment (zou3519) [reply test/test_meta.py:530]: Sorry, I am wrong about this. There is no test that runs meta tests over the entire pytorch codebase

File: test/test_meta.py [hunk 3]
```diff
@@ -1325,6 +1454,53 @@ def test_embedding_bag_byte_unpack(self):
             self.assertEqual(res.dtype, torch.float32)
             self.assertEqual(res.untyped_storage().data_ptr(), 0)
 
+    def test_meta_function_with_out(self):
```
Comment (lezcano) [inline test/test_meta.py:1457]: This test has now been superseded by the one above that tests on all PyTorch, so we can just remove it.

File: test/test_meta.py [hunk 4]
```diff
@@ -929,22 +947,86 @@ class MetaCrossRefDispatchMode(torch.utils._python_dispatch.TorchDispatchMode):
...
+            return (None, None, None)
+
+        candidate_ols = []
+        for candidate_ol_name in olp.overloads():
+            candidate_ol = getattr(olp, candidate_ol_name)
+            if any(arg.is_out for arg in candidate_ol._schema.arguments):
+                candidate_ols.append(candidate_ol)
+
+        if not candidate_ols:
+            MetaCrossRefDispatchMode.aten_olp_no_out_overload.add(olp)
```
Comment (lezcano) [inline test/test_meta.py:979]: why adding this before returning?

**Review by lezcano — COMMENTED** [review]
Uf, what a pain finding all those out variants. Just one question, but it generally LGTM

File: test/test_meta.py (hunk 4 above)
Comment (kundaMwiza) [reply test/test_meta.py:979]: This avoids trying to find an out overload for an `OpOverloadPacket` that is known not to have one from a previous search (Line 969 will return early)

**Review by lezcano — APPROVED** [review]
Thank you for the tests and the clean-up!

It very much looks like the issue with very many of these is that the line `autogen: fn.out` in `native_functions.yaml` does not expose the out parameter in Python (oops) (cc @bdhirsh who implemented quite a few codegen things)

A fantastic PR to this one would be to fix that and expose the out parameter for all those functions, in case you or @nicholasw-gc have the bandwith :)

**Comment by lezcano** [comment]
The failures are real though. Feel free to either submit fixes for those or xfail them (with comments describing the errors if possible :) )

File: test/test_meta.py [hunk 5]
```diff
@@ -968,6 +1058,45 @@ def __torch_dispatch__(self, func, types, args=(), kwargs=None):
...
+        # aten op overloads that have out parameters. Additionally, Python decompositions
+        # may register OpOverloadPacket's so decompositions need to be tested
+        # to ensure all OpOverloads still function for the Meta key (e.g. if a python decomposition
+        # is registered for an aten op aten.foo with overloads [default, out], the python
+        # function needs to support receiving `out` arguments)
+        if (
+            not self.inplace and
+            not self.supports_out and
+            (torch.is_tensor(expected) or isinstance(expected, Iterable))
+        ):
```
Comment (lezcano) [inline test/test_meta.py:1071]: Right. In that case, we should probably just test this if the default function is expected to pass, otherwise we end up testing the same xfail twice really. We have very little hope for the out test to pass when the default function does not pass.

**Comment by lezcano** [comment]
two side notes:
- It's not necessary to keep a clean git history, as PyTorch will squash all the commits before merging
- In later PRs, base them off `viable/strict`, as that's a branch whose tip should be always green. main may or may not be green on the whole CI (there's more CI than what's automatically run in each PR before merging)

**Comment by kundaMwiza** [comment]
@lezcano 
Not sure about the recent failures as I've updated this file: https://github.com/pytorch/pytorch/pull/107707/files#diff-2176e6e7254d7cca59eade91d31bb46a84e6c1f3f546892f47b0aeb0c418dceb

The decomp for `torch.full` is not compatible with `names` being provided as an argument so I've updated the above file.

**Comment by kundaMwiza** [comment]
Ah, Its a different test..
